Update dexter properties

daejunpark opened the pull request.

hjorthjort commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -5,11 +5,11 @@ Note that the real arithmetic is used throughout the document.
 - `lqtTotal == lqtAddress.totalSupply()`
 
-NOTE: `<` is possible when someone "donates" XTZ or tokens (i.e., sending funds without going through `add_liquidity()`).
+NOTE: `<` is possible when someone "donates" tokens (i.e., directly sending tokens to dexter without going through any dexter entrypoints).
```
Have we verified that it's impossible to send XTZ to a contract without triggering the `default` function? In Ethereum you can deposit funds into a contract without triggering a fallback, either by making the contract the `coinbase` of a block (and thus sending it block rewards), or by sending funds from a `selfdestruct` instruction.

daejunpark replied:
good point. @sskeirik do you have an idea?

hjorthjort commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -68,3 +68,1106 @@ where:
+
+syntax Int ::= Transfers(OpList) [function]
+rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) +Int T when From =/=K DEXTER andBool To ==K DEXTER
+rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) -Int T when From  ==K DEXTER
```
```suggestion
rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) -Int T when From  ==K DEXTER andBool To =/=K DEXTER
```

hjorthjort commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -68,3 +68,1106 @@ where:
+rule Sends(.List) => 0
+
+syntax Int ::= Transfers(OpList) [function]
+rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) +Int T when From =/=K DEXTER andBool To ==K DEXTER
+rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) -Int T when From  ==K DEXTER
```
```suggestion
rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) +Int T when From =/=K DEXTER andBool To  ==K DEXTER
rule Transfers([ Transaction DEXTER TOKEN 0 Transfer(From, To, T) ] ;; Ops) => Transfers(Ops) -Int T when From  ==K DEXTER andBool To =/=K DEXTER
```

daejunpark replied:
No, this covers the To == DEXTER case.

hjorthjort commented on tests/proofs/dexter/dexter-properties.md:
````diff
@@ -68,3 +68,1106 @@ where:
+
+```
+NOTE:
+- DEXTER may call XtzToToken or Default itself
````
```suggestion
- DEXTER may call XtzToToken or Default on itself
```

sskeirik commented on tests/proofs/dexter/dexter-properties.md:
````diff
@@ -1030,7 +1030,7 @@ proposition [dexter-emitted-ops]:
 ```
 
-The following proposition `[only-dexter]` states that no one other than Dexter can emit operations whose sender is Dexter.  Note that Sends, Transfers, and MintBurns are defined to be zero for operations whose sender is not Dexter.
+The following proposition `[only-dexter]` states that no one other than Dexter can emit operations whose sender is Dexter.  Note that Sends, Transfers, and MintBurns are defined to be zero for operations whose sender is not Dexter.  Also, obviously, the state variables and XTZ balance of Dexter can be updated only by Dexter.
````
The language that the "XTZ balance of Dexter can only be updated by Dexter" seems imprecise. Only a contract can _decrease_ its own XTZ balance --- because only it is able to send money from itself. However, conversely, only another contract can _increase_ that contract's own XTZ balance, because the money it receives must come from somewhere (excepting the case that the smart contract earns rewards through participating in the consensus process).

daejunpark replied:
good point. i slightly revised the text to better reflect my intention.

sskeirik commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -1174,30 +1194,33 @@ rule [is-valid]:
  andBool LqtMinted       ==Int XtzDeposited *Int L   /Int X
  andBool OpsEmitted ==K [ Transaction DEXTER TOKEN 0 Transfer(Sender, DEXTER, TokensDeposited) ]
                      ;; [ Transaction DEXTER LQT   0 Mint(Owner, LqtMinted) ]
+ andBool Sender =/=K DEXTER impliesBool B' ==Int B +Int XtzDeposited
+ andBool Sender  ==K DEXTER impliesBool B' ==Int B
```
Is it possible for Dexter to call AddLiquidity on itself? This seems inconsistent with comments made elsewhere.

daejunpark replied:
no it is not possible based on our assumption for the tezos execution model, but i'd like to make this rule to be independent from that.

sskeirik reviewed (commented):
Here are some initial comments.

sskeirik reviewed (approved):
Here are some more comments. I am going to approve since we want to merge this, even though there may be improvements that could be made.

daejunpark commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -1,70 +1,1542 @@
+The first invariant states that the Dexter state variables faithfully represent the actual pool reserves and liquidity supply.  That is, XtzPool and TokenPool must be equal to the actual XTZ and token reserves, and LqtTotal must be equal to the actual total liquidity supply.  Note that the Dexter entrypoint functions immediately update the state variables, while the actual reserves or supply are  ...
 
-### Reserve vs Liquidity
+The following claim `[inv-top-level]` states that the invariant holds at the completion of every top-level transaction.  Note that a top-level transaction is a transaction created by a non-contract account (i.e., a transaction whose sender is equal to the source), and the completion of a transaction involves the full execution “tree” following the DFS model adopted in the Florence upgrade.
```
i changed from "non-contract account" to "implicit account".

daejunpark commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -1,70 +1,1542 @@
+
+syntax Bool ::= TopLevelOps(OpList, Address) [function]
+rule TopLevelOps([ Transaction Sender _ _ _ ] ;; Ops, Source) => TopLevelOps(Ops, Source) ]] requires Sender ==K Source
+rule TopLevelOps([ Transaction Sender _ _ _ ] ;; _, Source) => false ]] requires Sender =/=K Source
```
fixed

daejunpark commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -1,70 +1,1542 @@
+- Note that Ops' must be of the form `Ops0 ;; [ Op ] ;; Ops1 ;; Ops2`, where Ops0 is possibly empty.  Since nothing in Ops' was emitted by Dexter, we have `Transfers(Ops') ==Int 0`.
+- By `[token-balance-of]`, we have `TokenPool ==Int D'`.
+- During the entire execution of Ops0 (including their nested sub-operations), `<tokenPool>` does not change, because no Dexter entrypoint can be executed due to the `<selfIsUpdatingTokenPool>` lock.  Thus, we have `T ==Int T'`.  (Note that, if the lock is somehow released during the execution of Ops0, then Op will fail which will revert the entire transaction.  Also, note that the lock cannot be ...
+- Also, during the execution of Ops0, `<tokenDexter>` cannot decrease, because only Dexter can spend its own tokens.  Thus, we have `D >=Int D'`.
```
it is indeed `D >= D'`.  but i guess the name `D'` may be confusing.  i renamed the variables.  please double-check.

daejunpark commented on tests/proofs/dexter/dexter-properties.md:
```diff
@@ -1,70 +1,1542 @@
+ andBool 0 <Int L' andBool L' ==Int S' +Int MintBurns(Ops' ;; Ops)
+
+proof [inv-update-token-pool-internal]:
+- apply [update-token-pool-internal]
+  - Amount ==Int 0 by assert
+  - Sender ==K TOKEN by assert
+- apply [lemma-token-pool-internal]
```
yes, you're right.  but i think it is fine for now.

daejunpark commented:
@sskeirik thanks for your review. i addressed all your comments.  please take a look.  i also put the `automerge` label.  i can improve further as separate prs.  thanks!
